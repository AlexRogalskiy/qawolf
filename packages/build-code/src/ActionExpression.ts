export class ActionExpression {
  code() {
    return "";
  }

  updatableCode() {
    return this.code();
  }
}
