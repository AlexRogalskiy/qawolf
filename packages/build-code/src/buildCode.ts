import { Workflow } from "@qawolf/types";
import { readFileSync } from "fs-extra";
import { compile } from "handlebars";
import { camelCase } from "lodash";
import { resolve } from "path";
import { formatLaunch } from "./formatLaunch";
import { formatStep } from "./formatStep";

type BuildCodeOptions = {
  test?: boolean;
  workflow: Workflow;
};

const scriptTemplate = compile(
  readFileSync(resolve(__dirname, "../static/script.hbs"), "utf8")
);

const testTemplate = compile(
  readFileSync(resolve(__dirname, "../static/test.hbs"), "utf8")
);

export const buildCode = (options: BuildCodeOptions) => {
  const { workflow } = options;

  const template = options.test ? testTemplate : scriptTemplate;

  const code = template({
    launch: formatLaunch(workflow.url, workflow.device),
    name: camelCase(workflow.name),
    steps: workflow.steps.map((step, i) => {
      const previousStep = i > 0 ? workflow.steps[i - 1] : null;

      return formatStep(step, previousStep);
    }),
    url: workflow.url
  });

  return code;
};
