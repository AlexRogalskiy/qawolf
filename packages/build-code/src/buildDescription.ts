import { Step } from "@qawolf/types";
import { describeDoc } from "@qawolf/web";
import { isNil } from "lodash";

export const buildDescription = (step: Step): string => {
  if (step.action === "scroll") {
    return `scroll`;
  }

  const description = describeDoc(step.html);

  const target = step.html.node;

  // link/input
  const tagName = `${target.name === "a" ? " link" : ` ${target.name}` || ""}`;

  if (step.action === "click") {
    return `click${description}${tagName}`;
  }

  if (step.action === "select") {
    return `select${description}`;
  }

  if (step.action === "type") {
    if (step.replace && (step.value === "" || isNil(step.value))) {
      return `clear${description}${tagName}`;
    }

    const value = step.value as string;
    if (value && value.indexOf("↓Enter") === 0) {
      return `Enter`;
    }

    if (value && value.indexOf("↓Tab") === 0) {
      return `Tab`;
    }

    return `type into${description}${tagName}`;
  }

  throw new Error(`Invalid step action ${step.action}`);
};
