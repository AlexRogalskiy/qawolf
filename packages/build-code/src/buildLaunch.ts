export const buildLaunch = (url: string, device?: string): string => {
  if (device) {
    return `await launch({ device: "${device}", url: "${url}" });`;
  }

  return `await launch({ url: "${url}" });`;
};
