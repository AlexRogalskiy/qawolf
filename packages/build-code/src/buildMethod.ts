import { ScrollValue, Step } from "@qawolf/types";
import { buildSelector } from "./buildSelector";

export const buildMethod = (step: Step, previousStep?: Step | null): string => {
  const selector = buildSelector(step);
  const options = buildMethodOptions(step, previousStep);

  const { action, value } = step;

  if (action === "click") {
    return `await browser.click(${selector}${options});`;
  }

  if (action === "scroll") {
    const scrollValue = value as ScrollValue;
    return `await browser.scroll(${selector}, { x: ${scrollValue.x}, y: ${scrollValue.y} }${options});`;
  }

  if (action === "select") {
    return `await browser.select(${selector}, ${JSON.stringify(
      value
    )}${options});`;
  }

  if (action === "type") {
    return `await browser.type(${selector}, ${JSON.stringify(
      value
    )}${options});`;
  }

  throw new Error(`Invalid step action ${action}`);
};

export const buildMethodOptions = (
  step: Step,
  previousStep?: Step | null
): string => {
  const pageChanged = previousStep && step.page !== previousStep.page;

  // we need to specify the page if the step has a different page
  // we don't need to, but we specify the page if it is not 0
  // for clarity and in case they delete steps
  if (pageChanged || step.page !== 0) {
    return `, { page: ${step.page} }`;
  }

  return "";
};
