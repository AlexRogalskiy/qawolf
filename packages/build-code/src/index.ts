export * from "./buildInitialCode";
export { stepToSelector } from "./buildSelector";
export * from "./buildVirtualCode";
export * from "./CodeReconciler";
