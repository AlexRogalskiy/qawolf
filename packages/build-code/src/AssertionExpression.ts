import { buildDescription } from "./buildDescription";
import { Expression } from "./Expression";
import { StepExpression } from "./StepExpression";

export class AssertionExpression implements Expression {
  private _stepExpression: StepExpression;

  constructor(stepExpression: StepExpression) {
    this._stepExpression = stepExpression;
  }

  private _formatIt() {
    const step = this._stepExpression.step();
    const description = buildDescription(step);
    if (!description.includes(`"`)) return `"can ${description}"`;
    if (!description.includes("'")) return `'can ${description}'`;
    return JSON.stringify("can " + description);
  }

  code() {
    return `\nit(${this._formatIt()}, async () => {\n  ${this._stepExpression.code()}\n});\n`;
  }

  updatableCode() {
    return this._stepExpression.updatableCode();
  }
}
