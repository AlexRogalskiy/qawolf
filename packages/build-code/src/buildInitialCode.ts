import { camelCase } from "lodash";
import { buildLaunch } from "./buildLaunch";

export type InitialCodeOptions = {
  device?: string;
  isTest?: boolean;
  name: string;
  patchHandle: string;
  url: string;
};

type TemplateOptions = {
  launch: string;
  name: string;
  patchHandle: string;
};

const buildInitialScript = ({ name, launch, patchHandle }: TemplateOptions) => {
  const code = `const { launch } = require("qawolf");
const selectors = require("../selectors/${name}");

const ${name} = async browser => {
  ${patchHandle}
};

exports.${name} = ${name};

if (require.main === module) {
  (async () => {
    const browser = ${launch}
    await ${name}(browser);
    await browser.close();
  })();
}`;

  return code;
};

const buildInitialTest = ({ name, launch, patchHandle }: TemplateOptions) => {
  const code = `const { launch } = require("qawolf");
const selectors = require("../selectors/${name}");

describe('${name}', () => {
  let browser;

  beforeAll(async () => {
    browser = ${launch}
  });

  afterAll(() => browser.close());
  ${patchHandle}
});`;

  return code;
};

export const buildInitialCode = (options: InitialCodeOptions) => {
  const templateOptions = {
    launch: buildLaunch(options.url, options.device),
    name: camelCase(options.name),
    patchHandle: options.patchHandle
  };

  if (options.isTest) {
    return buildInitialTest(templateOptions);
  }

  return buildInitialScript(templateOptions);
};
