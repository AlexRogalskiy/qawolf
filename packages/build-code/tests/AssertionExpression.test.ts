import { AssertionExpression } from "../src/AssertionExpression";
import { stepExpression } from "./fixtures";

describe("AssertionExpression.code()", () => {
  it("builds assertion code", () => {
    expect(new AssertionExpression(stepExpression).code()).toMatchSnapshot();
  });
});
