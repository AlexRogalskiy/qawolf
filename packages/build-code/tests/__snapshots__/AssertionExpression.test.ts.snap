// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssertionExpression.code() builds assertion code 1`] = `
"
it('can click \\"my-input\\" input', async () => {
  await browser.click({ css: \\"[data-qa='test-input']\\" });
});
"
`;
